src/main/java/com/gregtechceu/gtceu/integration/GTOreByProductWidget.java

```diff
@@ -127,24 +135,24 @@ public void setRecipe(GTOreByProduct recipeWrapper) {
                 xeiChance = chance.chance;
                 overlay = chance.createOverlay(false);
             }
-            if (itemOutputs.get(slotIndex).isEmpty() || itemOutputs.get(slotIndex).get(0).isEmpty()) {
+            if (itemOutputs.get(slotIndex).isEmpty()) {
                 itemOutputExists.add(false);
                 continue;
             }
 
             itemStackGroup.addWidget(new SlotWidget(itemOutputsHandler, slotIndex, ITEM_OUTPUT_LOCATIONS.get(i), ITEM_OUTPUT_LOCATIONS.get(i + 1))
                 .setCanTakeItems(false).setCanPutItems(false)
-                .setIngredientIO(IngredientIO.OUTPUT).setOverlay(overlay).setXEIChance(xeiChance)
+                .setIngredientIO(IngredientIO.OUTPUT).setXEIChance(xeiChance).setOverlay(overlay)
                 .setHoverTooltips(recipeWrapper.getTooltip(slotIndex + itemInputs.size())).setBackground((IGuiTexture) null));
             itemOutputExists.add(true);
         }
 
-        List<List<FluidStack>> fluidInputs = recipeWrapper.fluidInputs;
-        CycleFluidStorage fluidInputsHandler = new CycleFluidStorage(fluidInputs);
+        List<Either<List<Pair<TagKey<Fluid>, Long>>, List<FluidStack>>> fluidInputs = recipeWrapper.fluidInputs;
+        TagOrCycleFluidTransfer fluidInputsHandler = new TagOrCycleFluidTransfer(fluidInputs);
         WidgetGroup fluidStackGroup = new WidgetGroup();
         for (int i = 0; i < FLUID_LOCATIONS.size(); i += 2) {
             int slotIndex = i / 2;
-            if (!fluidInputs.get(slotIndex).isEmpty()) {
+            if (!fluidInputs.get(slotIndex).map(Function.identity(), Function.identity()).isEmpty()) {
```

Comment:
Is this by accident or necessary because of java generics having skill issues?

Reply:
maps an Either<List, List> to List